### PR TITLE
Add multiview support

### PR DESCRIPTION
**Pull request by LaylBongers**
## Connections
Fixes: https://github.com/gfx-rs/wgpu/issues/2186

## Description
Adds basic multiview support.

## Testing
You can now create a render pass with a target view with two array layers.

### CODE REVIEW

File: wgpu-core/src/command/render.rs [hunk 1]
```diff
@@ -154,6 +154,8 @@ pub struct RenderPassDescriptor<'a> {
     pub color_attachments: Cow<'a, [RenderPassColorAttachment]>,
     /// The depth and stencil attachment of the render pass, if any.
     pub depth_stencil_attachment: Option<&'a RenderPassDepthStencilAttachment>,
+    /// If this render pass will be rendering to multiple array layers.
+    pub multiview: bool,
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:158]: is this actually needed? If I understand correctly, we expect all of the attachments to be layered in this case, and with the same number of layers in them. So this flag becomes redundant - one could just check the actual texture views in the attachments.

File: wgpu-core/src/command/render.rs [hunk 2]
```diff
@@ -162,6 +164,7 @@ pub struct RenderPass {
     parent_id: id::CommandEncoderId,
     color_targets: ArrayVec<RenderPassColorAttachment, { hal::MAX_COLOR_TARGETS }>,
     depth_stencil_target: Option<RenderPassDepthStencilAttachment>,
+    multiview: bool,
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:167]: same here?

File: wgpu-core/src/pipeline.rs [hunk 3]
```diff
@@ -243,6 +243,8 @@ pub struct RenderPipelineDescriptor<'a> {
     pub multisample: wgt::MultisampleState,
     /// The fragment processing state for this pipeline.
     pub fragment: Option<FragmentState<'a>>,
+    /// If the pipeline will be used with a multiview render pass.
+    pub multiview: bool,
```
Comment (kvark) [inline wgpu-core/src/pipeline.rs:247]: would this need to know the number of layers?

File: wgpu-hal/src/vulkan/adapter.rs [hunk 4]
```diff
@@ -1052,6 +1052,7 @@ impl super::Adapter {
                 spv::Capability::ImageQuery,
                 spv::Capability::DerivativeControl,
                 spv::Capability::SampledCubeArray,
+                spv::Capability::MultiView,
```
Comment (kvark) [inline wgpu-hal/src/vulkan/adapter.rs:1055]: multi-view needs to be an optional feature exposed by the adapter and then activated by the user on device request.
The logic here, therefore, should be to only include this multi-view capability if the feature is activated.

**Review by kvark — CHANGES_REQUESTED** [review]
Looks great for a prototype!
In order to land this, we need a bit more polished look.
In addition to review notes, we'll also need to validate that multi-view is not used when not supported.

File: wgpu-core/src/pipeline.rs (hunk 3 above)
Comment (LaylBongers) [reply wgpu-core/src/pipeline.rs]: Tested with the mask set to all 1s, resulted in really bad driver crash on AMD so I added the layer count here.

File: wgpu-core/src/command/render.rs [hunk 5]
```diff
@@ -605,6 +614,41 @@ impl<'a, A: HalApi> RenderPassInfo<'a, A> {
...
 
+        let mut detected_mutiview: Option<Option<NonZeroU32>> = None;
+
+        let mut check_multiview = |view: &TextureView<A>| {
+            // Get the multiview configuration for this texture view
+            let layers = view.selector.layers.end - view.selector.layers.start;
+            let this_multiview = if layers == 1 {
+                None
+            } else {
+                let layers = NonZeroU32::new(layers).expect("Non-zero layers expected");
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:625]: this doesn't need to check for view layers being zero. View creation already does.
So this code could be just `NonZeroU32::new(layers)`

File: wgpu-core/src/command/render.rs [hunk 6]
```diff
@@ -605,6 +614,41 @@ impl<'a, A: HalApi> RenderPassInfo<'a, A> {
...
+                }
+            } else {
+                // Multiview is only supported if the feature is enabled
+                if this_multiview.is_some() && !device.features.contains(wgt::Features::MULTIVIEW) {
+                    return Err(RenderPassErrorInner::MissingFeatures(MissingFeatures(
+                        wgt::Features::MULTIVIEW,
+                    )));
+                }
+
+                detected_mutiview = Some(this_multiview);
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:647]: I think this logic isn't going to work right. It scans several attachments and only errors if we are switching from multiview to non-multiview. Imagine if the first analyzed attachment is non-layered but the second is layered. We should error, but this could wouldn't.

File: wgpu-core/src/command/render.rs [hunk 7]
```diff
@@ -745,6 +790,7 @@ impl<'a, A: HalApi> RenderPassInfo<'a, A> {
                 .views
                 .use_extend(&*view_guard, at.view, (), ())
                 .map_err(|_| RenderPassErrorInner::InvalidAttachment(at.view))?;
+            check_multiview(color_view)?;
             add_view(color_view, "color")?;
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:794]: we should do this logic in `add_view`, which is already called

File: wgpu-hal/src/vulkan/device.rs [hunk 8]
```diff
@@ -140,10 +142,30 @@ impl super::DeviceShared {
...
                     .subpasses(&vk_subpasses);
 
+                let mut multiview_info;
+                let mask;
+                if let Some(multiview) = e.key().multiview {
+                    // Sanity checks, better to panic here than cause a driver crash
+                    assert!(multiview.get() <= 8);
+                    assert!(multiview.get() > 1);
+
+                    let mut maskvalue = 0;
```
Comment (kvark) [inline wgpu-hal/src/vulkan/device.rs:156]: could be just `(1 << multiview.get())- 1`

File: wgpu-types/src/lib.rs [hunk 9]
```diff
@@ -526,6 +526,13 @@ bitflags::bitflags! {
         ///
         /// This is a native only feature.
         const SHADER_PRIMITIVE_INDEX = 1 << 39;
+        /// Enables multiview render passes and `builtin(view_index)` in vertex shaders.
+        ///
+        /// Supported platforms:
+        /// - Vulkan
+        ///
+        /// This is a native only feature.
+        const MULTIVIEW = 1 << 40;
```
Comment (kvark) [inline wgpu-types/src/lib.rs:535]: should this really be called `MULTIVIEW`? we don't have separate viewport and scissors for them (yet). It looks like this is more of `LAYERED_RENDERING` thing.

**Review by kvark — COMMENTED** [review]
Vulkan still needs the proper feature detection

File: wgpu-core/src/command/render.rs (hunk 6 above)
Comment (LaylBongers) [reply wgpu-core/src/command/render.rs:647]: The implemented logic here on every view is:
- check if multiview on or off has already been detected
- if it has: check if the multiview on/off and layer count matches
- if it hasn't: use the current view's value as the multiview detection

Subsequent checks will be checked against the first one's detected value. This is what the state of `detected_multiview` means:
- None: Not detected
- Some(None): Detected, no multiview
- Some(Some(2)): Detected, multiview with two layers

File: wgpu-core/src/command/render.rs (hunk 5 above)
Comment (LaylBongers) [reply wgpu-core/src/command/render.rs:625]: Did you mean `NonZeroU32::new_unchecked(layers)`?

File: wgpu-core/src/command/render.rs (hunk 7 above)
Comment (LaylBongers) [reply wgpu-core/src/command/render.rs:794]: `check_multiview` is called below on the resolve too, as it has to match, which does not have `add_view`. The function takes a mutable reference to `detected_multiview` and thus can't be mutable borrowed by `add_view` in addition to being used there.

File: wgpu-types/src/lib.rs (hunk 9 above)
Comment (LaylBongers) [reply wgpu-types/src/lib.rs:535]: This was discussed and resolved on the matrix chat, so I'll resolve this

File: wgpu-core/src/device/mod.rs [hunk 10]
```diff
@@ -2449,6 +2458,13 @@ impl<A: HalApi> Device<A> {
             .get(pipeline_layout_id)
             .map_err(|_| pipeline::CreateRenderPipelineError::InvalidLayout)?;
 
+        // Multiview is only supported if the feature is enabled
+        if desc.multiview.is_some() && !self.features.contains(wgt::Features::MULTIVIEW) {
```
Comment (kvark) [inline wgpu-core/src/device/mod.rs:2462]: we should use `self.require_features(..)` here

File: wgpu-core/src/command/render.rs [hunk 11]
```diff
@@ -605,6 +614,41 @@ impl<'a, A: HalApi> RenderPassInfo<'a, A> {
...
+            }
+
+            // Validate matching first, or store the first one
+            if let Some(multiview) = detected_multiview {
+                if multiview != this_multiview {
+                    return Err(RenderPassErrorInner::MultiViewMismatch);
+                }
+            } else {
+                // Multiview is only supported if the feature is enabled
+                if this_multiview.is_some() && !device.features.contains(wgt::Features::MULTIVIEW) {
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:641]: should use `device.require_features(..)` here

File: wgpu-hal/src/vulkan/adapter.rs [hunk 12]
```diff
@@ -23,6 +23,7 @@ pub struct PhysicalDeviceFeatures {
     image_robustness: Option<vk::PhysicalDeviceImageRobustnessFeaturesEXT>,
     robustness2: Option<vk::PhysicalDeviceRobustness2FeaturesEXT>,
     depth_clip_enable: Option<vk::PhysicalDeviceDepthClipEnableFeaturesEXT>,
+    multiview: Option<vk::PhysicalDeviceMultiviewFeatures>,
```
Comment (kvark) [inline wgpu-hal/src/vulkan/adapter.rs:26]: I think this should be `vulkan_1_1: vk::PhysicalDeviceVulkan11Features` and come right between `core` and `vulkan_1_2`.
In addition, I think we'd need `multiview: VkPhysicalDeviceMultiviewFeaturesKHR`, so that we can work with either Vk-1.1 or Vk-1.0 with extensions.

File: wgpu-hal/src/vulkan/adapter.rs [hunk 13]
```diff
@@ -1078,17 +1109,17 @@ impl super::Adapter {
                 lang_version: (1, 0),
                 flags,
                 capabilities: Some(capabilities.iter().cloned().collect()),
-                bounds_check_policies: naga::back::BoundsCheckPolicies {
-                    index: naga::back::BoundsCheckPolicy::Restrict,
+                bounds_check_policies: naga::proc::BoundsCheckPolicies {
+                    index: naga::proc::BoundsCheckPolicy::Restrict,
```
Comment (kvark) [inline wgpu-hal/src/vulkan/adapter.rs:1113]: ideally, we'd land this in a separate PR, but this isn't blocking

**Review by kvark — COMMENTED** [review]
This is shaping up well! I think it's almost ready, but we'll need one more iteration.
It would be nice to separate the Naga update from the rest, and we'll need to make Vulkan adapter logic a bit more powerful. I.e. it needs to support both extension and Vk_1_1 path for this, ideally.

File: wgpu-hal/src/vulkan/adapter.rs (hunk 13 above)
Comment (LaylBongers) [reply wgpu-hal/src/vulkan/adapter.rs:1130]: Decided to keep this in if it's not blocking, as the multiview builtin update is necessary for testing this functionality for me personally.

**Review by kvark — APPROVED** [review]
Great!
